langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java

```diff
@@ -98,7 +95,72 @@ static Iterable<JsonSchemaProperty> toJsonSchemaProperties(Parameter parameter)
             return removeNulls(STRING, enums((Object[]) type.getEnumConstants()), description);
         }
 
-        return removeNulls(OBJECT, description); // TODO provide internals
+        return removeNulls(OBJECT, schema(type), description);
+    }
+
+    private static JsonSchemaProperty schema(Class<?> structured){
+        //TODO: Externalize the depth
+        return schema(structured, new HashSet<>(), 3);
+    }
+
+    private static JsonSchemaProperty schema(Class<?> structured, Set<Class<?>> visited, int level) {
+        if (level < 0 || visited.contains(structured)) {
+            return null;
+        }
+
+        visited.add(structured);
+        Map<String,Object> properties = new HashMap<>();
+        for (Field field : structured.getDeclaredFields()) {
+            String name = field.getName();
+            if (name.equals("__$hits$__") || name.equals("this$0") || java.lang.reflect.Modifier.isStatic(field.getModifiers())) {
+                // Skip coverage instrumentation field.
+                continue;
+            }
+            Iterable<JsonSchemaProperty> schemaProperties = toJsonSchemaProperties(field, visited, level);
+            Map<Object,Object> objectMap = new HashMap<>();
+            for(JsonSchemaProperty jsonSchemaProperty : schemaProperties) {
+                objectMap.put(jsonSchemaProperty.key(), jsonSchemaProperty.value());
+            }
+            properties.put(name, objectMap);
+        }
+        return from( "schema", properties );
+    }
+
+    static Iterable<JsonSchemaProperty> toJsonSchemaProperties(Field field, Set<Class<?>> visited , int level) {
```

Comment:
there is a lot of duplicated logic copy-pasted from `toJsonSchemaProperties(Parameter parameter)`, can this be extracted and reused?

Reply:
Done